### PR TITLE
Add decoding datainterfaces

### PR DESCRIPTION
**Pull request by felixp8**
Add datainterfaces for RNN decoder logit output and LM decoded text output, for #5

### CODE REVIEW

**Comment by felixp8** [comment]
(no need to review just yet, still want to validate output a bit more)

**Comment by CodyCBakerPhD** [comment]
> (no need to review just yet, still want to validate output a bit more)

👍 Just ping me or mark 'ready for review' when ready then

**Comment by felixp8** [comment]
I think it's ready now, example file here: https://drive.google.com/drive/folders/1AlmK3W7uOwc6hWutmdT8Ln0IQ34RMle2?usp=sharing

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    ):
+        # initialize redis client and check connection
+        r = redis.Redis(
+            port=self.source_data["port"],
+            host=self.source_data["host"],
+        )
+        r.ping()
+
+        # get processing module
+        module_name = "decoding"
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: See [Best Practice for Module Names](https://nwbinspector.readthedocs.io/en/dev/best_practices/nwbfile_metadata.html#processing-module-names)

Since this is behavior related, would rename to `"behavior"`

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        module_description = "Contains decoder outputs for real-time speech decoding"
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]:
```suggestion
        module_description = "Contains decoder outputs for real-time speech decoding."
```

Description is fine even if you change to `"behavior"` - I do prefer though if they form complete sentences (i.e., punctuation, even if only a single sentence)

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        processing_module = get_module(nwbfile=nwbfile, name=module_name, description=module_description)
+
+        # get reference time for timestamps
+        # session_start_time = metadata["NWBFile"]["session_start_time"].timestamp()
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Commented line looks to be unused? Logically it seems fine to expect this in the metadata for an `add_to_nwbfile` step since it is a top-level required field to even create an `nwbfile`

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        timestamps = np.concatenate(timestamps)
+
+        # create timeseries obj
+        logits_timeseries = TimeSeries(
+            name="phoneme_logits",
+            data=logits,
+            unit="n.a.",
+            timestamps=timestamps,
+            description="Log-probabilities of the 39 phonemes plus silence and space between words, as "
+            + "predicted by an RNN decoder",
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Usual way of breaking this up stylistically with `black` is as follows

```suggestion
            description=(
                "Log-probabilities of the 39 phonemes plus silence and space between words, as "
                "predicted by an RNN decoder"
            ),
```

Comment (CodyCBakerPhD) [reply src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: triple-quotes are also accepted (stylistically and syntactically) but can look a little more awkward because the block wraps to the left edge after the first line

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            comments="The 41 columns correspond to, in order, silence, the space between words, and "
+            + "the 39 phonemes of the CMU pronouncing dictionary in alphabetical order",
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]:
```suggestion
            comments=(
                "The 41 columns correspond to, in order, silence, the space between words, and "
                "the 39 phonemes of the CMU pronouncing dictionary in alphabetical order"
            ),
```

similarly

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        # initialize redis client and check connection
+        r = redis.Redis(
+            port=self.source_data["port"],
+            host=self.source_data["host"],
+        )
+        r.ping()
+
+        # get processing module
+        module_name = "decoding"
+        module_description = "Contains decoder outputs for real-time speech decoding"
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Same comments about about module name/description

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        processing_module = get_module(nwbfile=nwbfile, name=module_name, description=module_description)
+
+        # get reference time for timestamps
+        # session_start_time = metadata["NWBFile"]["session_start_time"].timestamp()
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Same about metadata - I think we can trust the `session_start_time` being there, right?

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                    decoded_text.append(curr_text)
+                    decoded_timestamps.append(timestamp)
+                    last_text = curr_text
+            elif b"final_decoded_sentence" in entry[1].keys():  # reset on trial end
+                last_text = ""
+                # timestamp = int(entry[0].split(b'-')[0]) / 1000. - session_start_time
+                # # TODO: determine if this is necessary. does the model decide when to stop,
+                # # or is it externally controlled? if the former, then a STOP event might be useful
+                # decoded_text.append("STOP")
+                # decoded_timestamps.append(timestamp)
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py:148]: Have we asked about this yet?

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+        # get unique decoder outputs
+        output_set = sorted(list(set(decoded_text)))
+        decoded_text_idx = [output_set.index(sentence) for sentence in decoded_text]
+
+        # make labeledevents obj
+        events = LabeledEvents(
+            name="decoded_text",
+            description="Text decoded from RNN-predicted phonemes using language model",
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]:
```suggestion
            description="Text decoded from RNN-predicted phonemes using language model.",
```

**Comment by CodyCBakerPhD** [comment]
@felixp8 Overall looks really good!

Just minor comments with suggested changes; don't forget GitHub recently added a cool new batch commit feature if you decide to accept any of those

image

keeps the commit history deflated lol

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+            timestamps=decoded_timestamps,
+            data=decoded_text_idx,
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Rather minor, but curious how you think - I notice this ends up as a `int64` through the above steps (minimum native Python `int`) but casting as `int16` should suffice since I wouldn't expect there to be so many labels that we need to extend beyond that

But it's really not that much of a difference in data size either, so NBD either way

Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Could you compress it though? _I.e._, wrap in the `H5DataIO` with `"gzip"` default?

I realize that over the course of this project we will probably be undoing that whenever the NeuroConv Zarr refactor goes through, but we're not quite there yet and not sure how much longer the HDMF alterations will take

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+                trial_logits = []  # reset data stack
+            elif b"logits" in entry[1].keys():
+                trial_logits.append(entry)
+        logits = np.concatenate(logits, axis=0)
+        timestamps = np.concatenate(timestamps)
+
+        # create timeseries obj
+        logits_timeseries = TimeSeries(
+            name="phoneme_logits",
+            data=logits,
```
Comment (CodyCBakerPhD) [inline src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: Same comment as https://github.com/catalystneuro/stavisky-lab-to-nwb/pull/8/files#r1244282871 about compression

Can you also compress all the `timestamps` fields here too?

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py (hunk 3 above)
Comment (felixp8) [reply src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: yep, at some point I was testing these modules in isolation without running the whole NWBConverter so I was using a dummy `metadata` with no info. just forgot to switch back to using the metadata for the start time, will fix now

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py (hunk 10 above)
Comment (felixp8) [reply src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py]: yeah I'd expect `int16` should be fine. I'll cast to `int16` and just in case add a check right before to see if it'll overflow

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py (hunk 8 above)
Comment (CodyCBakerPhD) [reply src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py:148]: Just wondering - we can leave it here in the meantime as long as we don't forget by the end

**Review by CodyCBakerPhD — APPROVED** [review]

**Comment by CodyCBakerPhD** [comment]
LGTM!

File: src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py (hunk 8 above)
Comment (felixp8) [reply src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py:148]: ah whoops forgot to respond to this one. I believe the end of the trial is manually triggered with a button press, not by the model, which is why I commented out that code. But, I haven't asked them yet to make sure - didn't want to overwhelm them while they're working out timing issues. If a few more days pass without any updates from them, I will probably just ask anyway though so we're not sitting on this too long
